### PR TITLE
Support csproj files for identifying .NET Core projects

### LINKED ISSUE #11842: Support csproj files for identifying .NET Core projects
(This is a heads up for something that will become an issue in the future)

OpenShift supports .NET Core images and relies on `project.json` files to [identify a project as .NET](https://github.com/openshift/origin/blob/85eb37b34f0657631592356d020cef5a58470f8e/pkg/generate/source/detector.go#L71). An upcoming version of .NET Core will switch from using `project.json` files to `${PROJECT_NAME}.csproj` files: [.NET Core Tooling](https://blogs.msdn.microsoft.com/dotnet/2016/10/19/net-core-tooling-in-visual-studio-15/). This change is being added to Visual Studio right now and will be added to .NET Core in a later release (possibly in a couple of months).

OpenShift will need to update the detection routines to pick up these `csproj` files to identify a .NET Core project in addition to `project.json` files.

### PR DESCRIPTION
**Pull request by jim-minter**
Fixes #11842 

@bparees ptal - I removed DetectLiteralDotNet() because I couldn't see how it could possibly be triggered?

### CODE REVIEW

File: pkg/generate/source/detector.go [hunk 1]
```diff
@@ -29,70 +29,67 @@ var DefaultDetectors = Detectors{
...
 }
 
 // DetectDotNet detects .NET source and matches it to a dotnet supported annotatin or dotnet imagestream name
-func DetectDotNet(dir string) (*Info, bool) {
-	return detect("dotnet", dir, "project.json")
-}
-
-// DetectLiteralDotNet detects .NET source and matches it to a .net supported annotation
-func DetectLiteralDotNet(dir string) (*Info, bool) {
-	return detect(".net", dir, "project.json")
```
Comment (bparees) [inline pkg/generate/source/detector.go]: this matches to imagestreams with a "supports" annotation with value of ".net"

File: pkg/generate/source/detector.go [hunk 2]
```diff
@@ -29,70 +29,67 @@ var DefaultDetectors = Detectors{
...
 }
 
-func filesPresent(dir string, files []string) bool {
-	for _, f := range files {
-		_, err := os.Stat(filepath.Join(dir, f))
-		if err == nil {
-			return true
+func detectGlob(platform string, dir string, globs ...string) *Info {
+	for _, g := range globs {
+		if matches, _ := filepath.Glob(filepath.Join(dir, g)); len(matches) > 0 {
```
Comment (bparees) [inline pkg/generate/source/detector.go]: can we not just do globbing in the existing detect function, regardless of whether the passed pattern contains a * ?

**Comment by jim-minter** [comment]
All changes made, thanks.

File: pkg/generate/source/detector.go [hunk 3]
```diff
@@ -33,66 +30,56 @@ var DefaultDetectors = Detectors{
...
 	return detect("perl", dir, "index.pl", "cpanfile")
 }
 
 // DetectScala detects Scala source
-func DetectScala(dir string) (*Info, bool) {
+func DetectScala(dir string) *Info {
 	return detect("scala", dir, "build.sbt")
 }
 
 // DetectDotNet detects .NET source and matches it to a dotnet supported annotatin or dotnet imagestream name
```
Comment (bparees) [inline pkg/generate/source/detector.go]: annotation (i know, not your fault but it's bugging me now :) )

**Comment by jim-minter** [comment]
Done.

**Comment by bparees** [comment]
lgtm for post 3.4 merge.

**Comment by bparees** [comment]
[merge]

**Comment by bparees** [comment]
[merge]

**Comment by bparees** [comment]
```
# github.com/openshift/origin/pkg/generate/app
pkg/generate/app/strategyref_test.go:11: cannot use fakeDetector (type func(string) (*source.Info, bool)) as type source.DetectorFunc in array or slice literal
```

**Comment by bparees** [comment]
[merge]

**Pull request merged**